opa/pkg/main.go

```diff
@@ -0,0 +1,263 @@
+	if err != nil {
+		panic(err)
+	}
+	restMapper := restmapper.NewDiscoveryRESTMapper(groupResources)
+	// TODO filter by namespace
+	checkInstances, err := dynamicInterface.Resource(instanceGvr).Namespace("").List(ctx, metav1.ListOptions{})
+
+	if err != nil {
+		panic(err)
+	}
+	actionItems := make([]ActionItem, 0)
+
+	for _, checkInstance := range checkInstances.Items {
+		var checkInstanceObject customCheckInstance
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(checkInstance.Object, &checkInstanceObject)
+		if err != nil {
+			panic(err)
+		}
+		check, err := dynamicInterface.Resource(checkGvr).Namespace(checkInstanceObject.Namespace).Get(ctx, checkInstanceObject.Spec.CustomCheckName, metav1.GetOptions{})
+		if err != nil {
+			panic(err)
+		}
+		var checkObject customCheck
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(check.Object, &checkObject)
+		if err != nil {
+			panic(err)
+		}
+
+		newItems, err := processCheck(ctx, checkObject, checkInstanceObject, restMapper, dynamicInterface)
+		if err != nil {
+			panic(err)
+		}
+		actionItems = append(actionItems, newItems...)
+	}
+	outputFormat := Output{ActionItems: actionItems}
+	value, err := json.Marshal(outputFormat)
+	if err != nil {
+		panic(err)
+	}
+	err = ioutil.WriteFile(outputFile, value, 0644)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func processCheck(ctx context.Context, check customCheck, checkInstance customCheckInstance, restMapper meta.RESTMapper, dynamicInterface dynamic.Interface) ([]ActionItem, error) {
+	actionItems := make([]ActionItem, 0)
+
+	for _, target := range checkInstance.Spec.Targets {
+		for _, apiGroup := range target.ApiGroups {
+			for _, kind := range target.Kinds {
+				fmt.Printf("Starting to process %s %s\n\n", apiGroup, kind)
+				mapping, err := restMapper.RESTMapping(schema.GroupKind{Group: apiGroup, Kind: kind})
+				if err != nil {
+					return nil, err
+				}
+				gvr := mapping.Resource
+				list, err := dynamicInterface.Resource(gvr).Namespace("").List(ctx, metav1.ListOptions{})
+				if err != nil {
+					return nil, err
+				}
+				for _, obj := range list.Items {
+					query, err := rego.New(
+						rego.Query("results = data"),
+						rego.Module("fairwinds", check.Spec.Rego),
+						rego.Function1(
+							&rego.Function{
+								Name: "kubernetes",
+								Decl: types.NewFunction(types.Args(types.S), types.S),
+							},
+							func(_ rego.BuiltinContext, a *ast.Term) (*ast.Term, error) {
+
+								str, ok := a.Value.(ast.String)
+								if !ok {
+									return nil, nil
+								}
+								strValue := str.String()
+								if len(strValue) > 0 && strValue[0] == '"' {
+									strValue = strValue[1:]
+								}
+								if len(strValue) > 0 && strValue[len(strValue)-1] == '"' {
+									strValue = strValue[:len(strValue)-1]
+								}
+								var apiGroup string
+								for _, target := range check.Spec.AdditionalKubernetesData {
```

Comment:
we're getting pretty nested here 😅

Mind breaking out a separate function for the inside of this loop? Maybe the rego parts in particular.

Reply:
Yeah that nesting was getting out of control. I moved out the getKubernetesData function there